kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,750 @@
+    if not err then
+      local src = answers.errcode < CACHE_ONLY_EC and "server" or "client"
+      err = ("dns %s error: %s %s"):format(src, answers.errcode, answers.errstr)
+    end
+    table_insert(tries, { name .. ":" .. typstrs[qtype], err })
+  end
+
+  return answers, err
+end
+
+
+local function get_search_types(self, name, qtype)
+  local input_types = qtype and { qtype } or self.search_types
+  local checked_types = {}
+  local types = {}
+
+  for _, qtype in ipairs(input_types) do
+    if qtype == TYPE_LAST then
+      qtype = get_last_type(self.cache, name)
+    end
+    if qtype and not checked_types[qtype] then
+      table_insert(types, qtype)
+      checked_types[qtype] = true
+    end
+  end
+
+  return types
+end
+
+
+local function check_and_get_ip_answers(name)
+  if name:match("^%d+%.%d+%.%d+%.%d+$") then  -- IPv4
+    return {{ name = name, class = 1, type = TYPE_A, address = name }}
+  end
+
+  if name:match(":") then                     -- IPv6
+    return {{ name = name, class = 1, type = TYPE_AAAA, address = ipv6_bracket(name) }}
+  end
+
+  return nil
+end
+
+
+local function resolve_names_and_types(self, name, opts, tries)
+  local answers = check_and_get_ip_answers(name)
+  if answers then
+    answers.ttl = LONG_LASTING_TTL
+    answers.expire = now() + answers.ttl
+    return answers, nil, tries
+  end
+
+  -- TODO: For better performance, it may be necessary to rewrite it as an
+  --       iterative function.
+  local types = get_search_types(self, name, opts.qtype)
+  local names = search_names(name, self.resolv, self.hosts)
+
+  local err
+  for _, qtype in ipairs(types) do
+    for _, qname in ipairs(names) do
+      answers, err = resolve_name_type(self, qname, qtype, opts, tries)
+
+      -- severe error occurred
+      if not answers then
+        return nil, err, tries
+      end
+
+      if not answers.errcode then
+        insert_last_type(self.cache, qname, qtype) -- cache TYPE_LAST
+        return answers, nil, tries
+      end
+    end
+  end
+
+  -- not found in the search iteration
+  return nil, err, tries
+end
+
+
+local function resolve_all(self, name, opts, tries)
+  -- key like "short:example.com:all" or "short:example.com:5"
+  local key = "short:" .. name .. ":" .. (opts.qtype or "all")
+  opts.short_key = key  -- save for later use in the stale update task
+
+  stats_init(self.stats, name)
+  stats_count(self.stats, name, "runs")
+
+  if detect_recursion(opts, key) then
+    stats_count(self.stats, name, "fail_recur")
+    return nil, "recursion detected for name: " .. name
+  end
+
+  -- quickly lookup with the key "short:<name>:all" or "short:<name>:<qtype>"
+  local answers, err, hit_level = self.cache:get(key)
+  if not answers then
+    answers, err, tries = resolve_names_and_types(self, name, opts, tries)
+    if not opts.cache_only and answers then
+      -- insert via the `:get` callback to prevent inter-process communication
+      self.cache:get(key, nil, function()
```

Comment:
If another worker resolved the name between the last `self.cache:get` and this `self.cache:get`, then it is okay? Could you comment and explicitly mention this?